doc/design/doi-generation-2.md

```diff
@@ -0,0 +1,212 @@
+
+DataCite allows for three types of DOIs ([DataCite](https://support.datacite.org/docs/what-does-the-state-of-the-doi-mean)):
+
+- `Draft`. We do not use those.
+  *Can be deleted, and they require only the identifier itself in order to be created or saved. They can be updated to either Registered or Findable DOIs. Registered and Findable DOIs may not be returned to the Draft state, which means that changing the state of a Draft is final.*
+- `Registered`. Like `Findable` but not indexed for search, so we do not use them.
+- `Findable`. Is the type we use for published dandisets.
+  Requires to be valid (pass validation to fit the datacite schema) to be created.
+
+We propose to:
+
+- Upon creation of a **public** dandiset, mint a `Dandiset DOI` (a DataCite `Draft DOI`) `10.48324/dandi.{dandiset.id}` with
+  - *minimal metadata* entered during creation request (title, description, license)
+  - DLP URL `https://dandiarchive.org/dandiset/{dandiset.id}`
+  - For embargoed dandiset, **do not** specify any metadata besides the DLP URL.
+  - If minting a DOI fails, we need to raise exception to inform developers about the issue but proceed with the creation of the dandiset.
+- Upon changes to a non-embargoed, draft dandiset metadata record:
+  - If `Draft DOI`, attempt to "promote" it to `Findable`.
+    - If validation fails - keep `Draft DOI` (very limited validation), attempt to update datacite metadata record while keeping the same target URL.
+    - **Question to clear up**: what happens to `Draft DOI` if metadata record is invalid? It seems to create one with no metadata, but does it update only the fields it knows about?
+  - If `Findable DOI`
+    - if draft version with no prior publications, but which had legit metadata, we try to update metadata.
+      - **Question to clear up** If fails, we either ignore or just add a comment somewhere that "record might not reflect the most recent changes to draft version".
+      - **Question to clear up** If we add to validation procedures to dandiset updates, (validation against datacite metadata record), we can report errors to the user so they can be addressed prior to attempted publication. May be we should validate only if no other errors (our schema validation) were detected to reduce noise, or just give a summary that "Metadata is not satisfying datacite model, fix known metadata errors first."
+    - if dandiset was published at least once (has version) -- we do not update anything since DLP points to that published version.
+  - **TODO: figure out how to annotate Draft version, so it always says that it is a draft version and thus potentially not used for citation if that could be avoided**
+- Upon changes to embargoed dandiset metadata record, don't do anything.
+- Upon unembargoing dandiset: update `Draft DOI` metadata record with current metadata and attempt to promote to a `Findable DOI` **after** unembargoing.
```

Comment:
There is a small difference here with the diagram. Can we safely assume that an embargoed dataset has any DOI at all? This may boil down to the question "Can a dandiset be embargoed at creation time"

Assuming that creating an embargoed dandiset does NOT get a DOI, but creating a dandiset first, then embargoing it are both possibilities, at unembargo time we just check to see if theres a DOI or not, and proceed accordingly.

Reply:
An alternative workflow would be to create a draft DOI for embargoed datasets, and attempt to promote it on each update, but instead of promoting to Findable, we promote to Registered (which is hidden from public).

We can "publish" Registered DOIs which become Findable when unembargoed.

Reply:
> This may boil down to the question "Can a dandiset be embargoed at creation time"

This is the only way to get dandiset embargoed -- at creation time. And as a part of this design we want to avoid adding any fake DOIs as it is done now.

> An alternative workflow would be to create a draft DOI for embargoed datasets, and attempt to promote it on each update, but instead of promoting to Findable, we promote to Registered (which is hidden from public).

It is a sound alternative but ATM we are not exposing metadata of embargoed dandisets at all. If we expose it in Registered, it would be a "side effect"... so let's keep it just not populated at all.

Reply:
And otherwise you didn't spot any other divergence in the diagram? (I am yet to look too)

Reply:
I need to have another careful look-- after "first contact with reality", ie playing with the code, this does need to change a bit.